src/librustc_typeck/check/mod.rs

```diff
@@ -3093,14 +3093,31 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             !error_happened &&
             !remaining_fields.is_empty()
         {
-            span_err!(tcx.sess, span, E0063,
-                      "missing field{} {} in initializer of `{}`",
-                      if remaining_fields.len() == 1 {""} else {"s"},
-                      remaining_fields.keys()
-                                      .map(|n| format!("`{}`", n))
-                                      .collect::<Vec<_>>()
-                                      .join(", "),
-                      adt_ty);
+            let truncated_fields = match remaining_fields.len() {
+                len if len <= 3 => (remaining_fields.keys().take(len), "".to_string()),
+                len => {
+                    let x = len - 3;
```

Comment:
I'd be tempted to use an if-expression instead.

```rust
let len = remaining_fields.len();
let truncated_fields = if len <= 3 {
        (remaining_fields.keys().take(len), "".to_string())
    } else {
        (remaining_fields.keys().take(3), format!(", and {} other field{}",
                            (len-3), if x == (len-3) {""} else {"s"}))
    };
```

There are probably are slicker fixes, but maybe that's an idea